### PR TITLE
Fix HMR failing when rendering with `/index`

### LINKED ISSUE #6497: HMR Not working for Clean URLs (when redirecting to /pages/index.js)
# Bug report

HMR is not working for custom URL. If you redirect to index it will fail, if you redirect to another file, it will work.

## To Reproduce

1. run `npm run dev`
2. update something in /pages/index.js file
3. next compiles but browser doesn't update

## Expected behavior

Browser updates

## System information
 - OS: [e.g. macOS]
 - Browser: chrome 
 - Version of Next.js: 8.0.3

## Additional context

server.js file:

```
const express = require('express');
const next = require('next');

const dev = process.env.NODE_ENV !== 'production';
const app = next({ dev });
const handle = app.getRequestHandler();

app.prepare()
  .then(() => {
    const server = express();

    server.get('/test', (req, res) => {
      app.render(req, res, '/index', {});
    });

    server.get('*', handle);

    server.listen(3000, (err) => {
      if (err) throw err;
      console.log('> Ready on http://localhost:3000')
    })
  })
  .catch((ex) => {
    console.error(ex.stack);
    process.exit(1)
  });
```

/pages/index.js:

```
import React, { Component } from 'react';

class Index extends Component {
 render() {
   return (
     <div>
       <p>Hello Next.js</p>
     </div>
   )
 }
}

export default Index;

```

### PR DESCRIPTION
**Pull request by ijjk**
Fixes: #6497

### CODE REVIEW

File: packages/next/server/next-dev-server.js [hunk 1]
```diff
@@ -108,7 +108,7 @@ export default class DevServer extends Server {
       }
       if (!this.quiet) console.error(err)
     }
-
+    pathname = pathname === '/index' ? '/' : pathname
```
Comment (timneutkens) [inline packages/next/server/next-dev-server.js]: I wonder if we should do this in render.tsx instead 🤔

Comment (ijjk) [reply packages/next/server/next-dev-server.js]: That works 👍

**Review by Timer — COMMENTED** [review]
What happens for other routes? e.g. `/docs/index`? Do we need to handle those cases?

TBH I feel like `/index` should 404 -- not resolve to a page.

File: packages/next-server/server/next-server.ts [hunk 2]
```diff
@@ -285,6 +285,7 @@ export default class Server {
     query: ParsedUrlQuery = {},
     opts: any,
   ) {
+    pathname = pathname.replace(/\/index/, '') || '/'
```
Comment (Timer) [inline packages/next-server/server/next-server.ts]: Wait, what happens if the path is like, `/indexer`? Should you anchor this?

Comment (timneutkens) [reply packages/next-server/server/next-server.ts]: `/\/index/$`

**Review by timneutkens — APPROVED** [review]

**Pull request merged**